common/src/main/java/io/crate/types/TypeSignatures.java

```diff
@@ -0,0 +1,214 @@
+        String base = signature.getBase();
+        List<TypeSignatureParameter> parameters = signature.getParameters();
+        if (base.equalsIgnoreCase(ArrayType.NAME)) {
+            if (parameters.size() == 0) {
+                return new ArrayType<>(UndefinedType.INSTANCE);
+            }
+            DataType<?> innerType = getType(parameters.get(0).getTypeSignature());
+            return new ArrayType<>(innerType);
+        }
+        if (base.equalsIgnoreCase(ObjectType.NAME)) {
+            var builder = ObjectType.builder();
+            for (int i = 0; i < parameters.size() - 1;) {
+                var valTypeSignature = parameters.get(i + 1);
+                builder.setInnerType(String.valueOf(i), getType(valTypeSignature.getTypeSignature()));
+                i += 2;
+            }
+            return builder.build();
+        }
+        return DataTypes.ofName(signature.getBase());
+    }
+
+    @Nullable
+    public static DataType<?> getCommonSuperType(DataType<?> firstType, DataType<?> secondType) {
+        TypeCompatibility compatibility = compatibility(firstType, secondType);
+        if (!compatibility.isCompatible()) {
+            return null;
+        }
+        return compatibility.getCommonSuperType();
+    }
+
+    public static boolean canCoerce(DataType<?> fromType, DataType<?> toType) {
+        return fromType.isConvertableTo(toType);
+    }
+
+    private static TypeCompatibility compatibility(DataType<?> fromType, DataType<?> toType) {
+        if (fromType.equals(toType)) {
+            return TypeCompatibility.compatible(toType, true);
+        }
+
+        if (fromType.equals(UndefinedType.INSTANCE)) {
+            return TypeCompatibility.compatible(toType, true);
+        }
+
+        if (toType.equals(UndefinedType.INSTANCE)) {
+            return TypeCompatibility.compatible(fromType, false);
+        }
+
+        // If given types share the same base, e.g. arrays, parameter types must be compatible.
+        String fromTypeBaseName = fromType.getTypeSignature().getBase();
+        String toTypeBaseName = toType.getTypeSignature().getBase();
+        if (fromTypeBaseName.equals(toTypeBaseName)) {
+            if (isCovariantParametrizedType(fromType)) {
+                return typeCompatibilityForCovariantParametrizedType(fromType, toType);
+            }
+            return TypeCompatibility.compatible(fromType, false);
+        }
+
+        // Use possible common super type (safe conversion)
+        DataType<?> commonSuperType = convertTypeByPrecedence(fromType, toType);
+        if (commonSuperType != null) {
+            return TypeCompatibility.compatible(commonSuperType, commonSuperType.equals(toType));
+        }
+
+        // Try to force conversion, first to the target type or if fails to the source type (possible unsafe conversion)
+        DataType<?> coercedType = coerceTypeBase(fromType, toType.getTypeSignature().getBase());
```

Comment:
Would it make sense to have the `getBase()`  directly in `DataType` ? E.g. `type.getBaseName()`. Would make sure that we don't construct the `TypeSignature` multiple times.